### PR TITLE
[Relayer] feat: Add Relayer struct

### PR DESCRIPTION
**Pull request by red-0ne**
### Summary generated by Reviewpad on 10 Nov 23 22:33 UTC
This pull request includes the following changes:

- In `pkg/relayer/interface.go`, the import statement for `github.com/pokt-network/poktroll/x/service/types` has been removed and replaced with `types`.
- In `pkg/relayer/proxy/proxy.go`, the comment for the `Start` method has been updated to clarify that it is blocking as long as all `RelayServers` are running.
- A new file `pkg/relayer/relayminer.go` has been added, which contains a new struct `relayMiner` and its corresponding methods.
- In `pkg/relayer/session/session.go`, there is a typo fix in a comment, changing "attampting" to "attempting".

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: pkg/relayer/relayer.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"context"
+
+	"cosmossdk.io/depinject"
+)
+
+type RelayerOption func(*Relayer)
+
+// Relayer is the main struct that encapsulates the relayer's responsibilities.
```
Comment (Olshansk) [inline pkg/relayer/relayer.go:11]:
```suggestion
// Relayer is the main struct that encapsulates the relayer's responsibilities (i.e. Relay Mining).
```

File: pkg/relayer/relayer.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+// It starts and stops the RelayerProxy and provide the served relays observable to them miner.
+type Relayer struct {
```
Comment (Olshansk) [inline pkg/relayer/relayer.go:13]: Thoughts on calling this struct (and the file) `RelayMiner`? @bryanchriswhite

File: pkg/relayer/relayer.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	return rel, nil
+}
+
+// Start provides the miner with the served relays observable and starts the relayer proxy.
+// This method is blocking while the relayer proxy is running and returns when Stop is called
+// or when the relayer proxy fails to start.
+func (rel *Relayer) Start(ctx context.Context) error {
+	rel.miner.MineRelays(ctx, rel.relayerProxy.ServedRelays())
```
Comment (Olshansk) [inline pkg/relayer/relayer.go:41]: Not clear if this is a background go routine or not. I think it means we either have to add comments and/or rename downstream functions.

maybe `rel.miner.StartMining`?

File: pkg/relayer/relayer.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	return rel.relayerProxy.Start(ctx)
+}
+
+// Stop stops the relayer proxy which in turn stops all advertised relay servers
+// and unsubscribes the miner from the served relays observable.
+func (rel *Relayer) Stop(ctx context.Context) error {
+	return rel.relayerProxy.Stop(ctx)
```
Comment (Olshansk) [inline pkg/relayer/relayer.go:48]: Do we have to stop mining?

File: pkg/relayer/relayer.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+	return rel, nil
+}
+
+// Start provides the miner with the served relays observable and starts the relayer proxy.
+// This method is blocking while the relayer proxy is running and returns when Stop is called
+// or when the relayer proxy fails to start.
+func (rel *Relayer) Start(ctx context.Context) error {
+	rel.miner.MineRelays(ctx, rel.relayerProxy.ServedRelays())
+	return rel.relayerProxy.Start(ctx)
```
Comment (Olshansk) [inline pkg/relayer/relayer.go:42]: This might involve some downstream dependencies, but I think it'd be "c

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: pkg/relayer/relayer.go (hunk 3 above)
Comment (red-0ne) [reply pkg/relayer/relayer.go:41]: Not a background go routine.
Add comments: Yes.
Rename to `StartMining[Relays]`: Yes
:+1:

File: pkg/relayer/relayer.go (hunk 4 above)
Comment (red-0ne) [reply pkg/relayer/relayer.go:48]: I need to double check, but the idea is that stopping the proxy would unsubscribe so mining stops automatically.

File: pkg/relayer/relayer.go (hunk 5 above)
Comment (red-0ne) [reply pkg/relayer/relayer.go:42]: Calling `Start` should be enough

File: pkg/relayer/relayer.go (hunk 2 above)
Comment (bryanchriswhite) [reply pkg/relayer/relayer.go:13]: :joy: I would love that - haha. That's what we originally wanted to call it but IIRC there was some sort of decree that was made and "it shall be called 'the relayer'". 

To be clear on file naming, I would propose `relay_miner.go`.

File: pkg/relayer/relayer.go (hunk 3 above)
Comment (bryanchriswhite) [reply pkg/relayer/relayer.go:41]: `#MineRelays()` will not block:

```
// MineRelays kicks off relay mining by mapping the servedRelays ovservable through
// a pipeline which hashes the relay, checks if it's above the mining difficulty,
// adds it to the session tree, and then maps any errors to a  new observable.
// It also starts the claim and proof pipelines which are subsequently driven by
// mapping over RelayerSessionsManager's SessionsToClaim return observable.
// It does not block as map operations run in their own goroutines.
```

**Comment by red-0ne** [comment]
We'll wait for #168 to be merge to main before this could be merged

File: pkg/relayer/relayminer.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+import (
+	"context"
+	"log"
+
+	"cosmossdk.io/depinject"
+)
+
+// relayMiner is the main struct that encapsulates the relayer's responsibilities (i.e. Relay Mining).
+// It starts and stops the RelayerProxy and provide the served relays observable to them miner.
```
Comment (Olshansk) [inline pkg/relayer/relayminer.go:11]:
```suggestion
// It starts and stops the RelayerProxy and provide the served relays observable to the miner.
```

File: pkg/relayer/relayminer.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+// This method is blocking while the relayer proxy is running and returns when Stop is called
+// or when the relayer proxy fails to start.
+func (rel *relayMiner) Start(ctx context.Context) error {
+	// relayerSessionsManager.Start does not block.
+	// Set up the session (proof/claim) lifecycle pipeline.
+	log.Println("INFO: Starting relayer sessions manager...")
+	rel.relayerSessionsManager.Start(ctx)
+
+	// Start the flow of relays by starting relayer proxy.
+	// This is a blocking call as it waits for the waitgroup to be done.
```
Comment (Olshansk) [inline pkg/relayer/relayminer.go:50]: Which waitgroup are you referring to?

File: pkg/relayer/relayminer.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	log.Println("INFO: Starting relayer proxy...")
+	if err := rel.relayerProxy.Start(ctx); err != nil {
```
Comment (Olshansk) [inline pkg/relayer/relayminer.go:52]: Is blocking inside of `Start` something we want to keep around long-term or just a current solution? Fine either way but just asking to understand where your mind is at.

Might be a `TODO_REVISIT` if you think it's worth it but not a blocker!

**Review by Olshansk — APPROVED** [review]
A couple small comments & questions, but no blockers!

File: pkg/relayer/relayminer.go (hunk 7 above)
Comment (red-0ne) [reply pkg/relayer/relayminer.go:50]: Silly copy past from #177 :facepalm:

File: pkg/relayer/relayminer.go (hunk 8 above)
Comment (red-0ne) [reply pkg/relayer/relayminer.go:52]: I think we want to keep `Start` blocking as it's starting the relayer proxies. The passed context (from cmd) could close it anyway. And this is the top level function that is called from cmd. cc @bryanchriswhite

Comment (Olshansk) [reply pkg/relayer/relayminer.go:52]: Sgtm!

**Review by Olshansk — APPROVED** [review]

**Pull request merged**